UOP1_Project/Assets/Scripts/Interaction/InteractionManager.cs

```diff
@@ -109,44 +106,63 @@ void OnInteractionButtonPress()
 
 	private void OnTriggerEnter(Collider other)
 	{
+		InteractionType ongoingInteractionType = InteractionType.None;
+
 		if (other.CompareTag("Pickable"))
 		{
-			_potentialInteraction = InteractionType.PickUp;
+			ongoingInteractionType = InteractionType.PickUp;
 			//Debug.Log("I triggered a pickable object!");
-			DisplayInteractionUI();
 		}
 		else if (other.CompareTag("CookingPot"))
 		{
-			_potentialInteraction = InteractionType.Cook;
+			ongoingInteractionType = InteractionType.Cook;
 			//Debug.Log("I triggered a cooking pot!");
-			DisplayInteractionUI();
 		}
 		else if (other.CompareTag("NPC"))
 		{
-			_potentialInteraction = InteractionType.Talk;
+			ongoingInteractionType = InteractionType.Talk;
 			//Debug.Log("I triggered an NPC!");
+		}
+		if (ongoingInteractionType != InteractionType.None)
+		{
+			_ongoingInteractions.Push(new Interaction(ongoingInteractionType, other.gameObject));
 			DisplayInteractionUI();
 		}
-		_currentInteractableObject = other.gameObject;
 	}
 
 	private void DisplayInteractionUI()
 	{
 		//Raise event to display UI
-		_toggleInteractionUI.RaiseEvent(true, _potentialInteraction);
+		Interaction onGoingInteraction = _ongoingInteractions.Peek();
+		_toggleInteractionUI.RaiseEvent(true, onGoingInteraction.Type);
 	}
 
 	private void OnTriggerExit(Collider other)
 	{
-		ResetInteraction();
+		ResetInteraction(other.gameObject);
 	}
 
-	private void ResetInteraction()
+	private void ResetInteraction(GameObject obj)
 	{
-		_potentialInteraction = InteractionType.None;
-		_currentInteractableObject = null;
+		Stack<Interaction> updatedStack = new Stack<Interaction>();
+		while(_ongoingInteractions.Count > 0)
+		{
+			Interaction interaction = _ongoingInteractions.Pop();
+			if (interaction.InteractableObject != obj)
+			{
+				updatedStack.Push(interaction);
+			}
+		}
+		_ongoingInteractions = updatedStack;
+
+		if (_ongoingInteractions.Count > 0)
+		{
+			_toggleInteractionUI.RaiseEvent(true, _ongoingInteractions.Peek().Type);
 
-		if (_toggleInteractionUI != null)
-			_toggleInteractionUI.RaiseEvent(false, _potentialInteraction);
+		}
+		else
+		{
+			_toggleInteractionUI.RaiseEvent(false, InteractionType.None);
+		}
```

Comment:
How about if instead of creating new stack, you'd iterate over _ongoingInteractions and remove whatever interaction we've found. If the interaction we've found was the most recent one (the one we are currently interacting with) AND stack is not empty you call DisplayInteractionUI, othherwise _toggleInteractionUI.RaiseEvent(false, InteractionType.None);

This should be more optimal and slightly cleaner as you reuse DisplayInteraction method.

Reply:
Actually, I did some searches and it seems not possible to delete an item in the middle of a stack in C# without having to create a new one. I am not "fluent" in C# so if you have a working proposal I will be happy to integrate it.

For the last part of your comment, it is done on purpose, when an interactable item has been consumed (pick up item), the collider is removed so we have to show the InteractionUI for the next one. Next indeed, in case of random collider exit (not related to interactable objects), the current implementation will inverse the interaction order and so propose another interaction. Not sure it is a problem, since all are valid, most of the case it might be the same type, the other cases might be rare (ie: pickup-able item close to an NPC and the player playing the attack action which will trigger a random collider exit with the cane)

Reply:
I have updated the implementation to use a LinkedList instead of a Stack. It provides naturally remove item from the collection and simple accessor to get/remove first item.